### ISSUE TITLE
Retrieve vector values using cartesian coordinates

### ISSUE THREAD

**Issue by Kaesebrot84** [toji/gl-matrix#276]
After using `gl-matrix` for a while I found it a little annoying that I have to call vectors components using array notation like for the `x` value of a `vec3`:

```javascript
let vector = glMatrix.vec3.fromValues(0,0,0);
// For the x value I call
let x = vector[0];
```

Please correct me if I am wrong, but wouldn't it be convenient if we could call the single vector components by their cartesian names? I am thinking of something like:

```javascript
let vector = glMatrix.vec3.fromValues(0,0,0);
// For the x value I call
let x = vector.x;
```

Are there any reasons why there is no such option? I would be happy to volunteer for such task.

**Comment by stefnotch** [comment]
Yes, it would be very convenient!

However, to do that, you would either have to:
- extend the native Float32Array: Not necessarily a good idea
- return a custom object: Not exactly great for performance. (I think) Also, I don't think that you can easily pass that custom object to functions like [gl.bufferData](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/bufferData)

If you can figure out how to do it while keeping this library fast, it would be great! :)

**Comment by Kaesebrot84** [comment]
@stefnotch You are right, I had something different in mind though concerning the following point:

> return a custom object: Not exactly great for performance. (I think) Also, I don't think that you can easily pass that custom object to functions like gl.bufferData

The idea was to add the option of using cartesian properties not to entirely omit the array notation.

In my project I am using the [@types/gl-matrix](https://www.npmjs.com/package/@types/gl-matrix), in which case it would certainly be convenient to use cartesian properties. Or would it be possible to make this happen for @types even if it is not possible to use in vanilla JS?

Anyway I will try to think of a concept.

**Comment by stefnotch** [comment]
@Kaesebrot84 
Well, if you decide to create an object with extends/inherits from Float32Array, it would be a custom/not-native object.

I have never worked with @types, so, I have no clue if this is possible. However, you would probably need a way to distinguish between normal Float32Arrays and the Float32Arrays which are returned by gl-matrix. So, you most likely need a custom object?


Note: gl-matrix returns normal arrays, if Float32Arrays aren't supported

**Comment by stefnotch** [comment]
@Kaesebrot84 Have you figured out anything so far?

**Comment by Kaesebrot84** [comment]
@stefnotch Sorry I have been busy over the holidays. I will look into as soon as I have time though.

**Comment by Kaesebrot84** [comment]
@stefnotch I had a look at the example you posted above and it seems to aim at something like:

```javascript
class Cartesian extends Array {
	constructor(...args) {
            super(...args);
	}
	
	x() { return this[0]; }	
	y() { return this[1]; }	
	z() { return this[2]; }
}

let myArray = new Cartesian(1,2,3);
console.log('myArray: ', myArray);
console.log('Type of myArray: ', typeof myArray);
console.log('Is array: ', myArray instanceof Array);
console.log('x:', myArray.x());
console.log('y:', myArray.y());
console.log('z:', myArray.z());
```

This does work as expected. I just want to make clear that we have the same idea. Next I would:

1. Make `vec2`, `vec3`, `vec4` etc.  extend `Float32Array | Array` and implement all the needed functions in it.

1.  Somehow make sure that I am actually using the native `Float32Array` if possible.

1. Create a jsperf benchmark.

Am I missing something?

**Comment by stefnotch** [comment]
@Kaesebrot84 Oh, great to see some example code! Looks like it works really well.

Though, after testing a bit on Firefox, it looks like creating a non-native object does incur quite a performance penalty.
```
let x = new Array(Math.floor(Math.random() * 100));
let y = new Cartesian1(Math.floor(Math.random() * 100)); //Runs a lot slower
```
Also, accessing the elements using their index does not incur a statistically significant performance penalty.

Because this library is mostly intended to be used with WebGL and should be as fast as possible, chances are high that @toji will reject this idea.

Though, if you just want to have those convenience functions in a fork or in your own project, all you would have to change is this: https://github.com/toji/gl-matrix/blob/master/src/gl-matrix/common.js#L28
(Here is the relevant piece of code in the compiled version of glmatrix https://github.com/toji/gl-matrix/blob/master/dist/gl-matrix.js#L143 )
e.g.
```
class Cartesian extends Array {
	constructor(...args) {
            super(...args);
	}
	
	get x() { return this[0]; }	
	get y() { return this[1]; }	
	get z() { return this[2]; }
	set x(val) { this[0] = val; }	
	set y(val) { this[1] = val; }	
	set z(val) { this[2] = val; }
}
export let ARRAY_TYPE = Cartesian;
```

**Comment by Kaesebrot84** [comment]
@stefnotch Looks nice I have not tried it yet, tbh. But I will.
